### PR TITLE
Implement orientation (placement of master area) for master layout

### PR DESCRIPTION
**Pull request by proycon**
#### Describe your PR, what does it fix/add?

This implements the idea I posted in https://github.com/hyprwm/Hyprland/issues/1059 to make the master layout more friendly for use on vertical monitors. Each workspace now has an `orientation` which states where the master area should be (I opted for the more generic term `orientation` rather than `masterplacement`).  The values can be `left` (default, so nothing changes for existing configurations), `top`, `right` and `bottom`.

For the layoutmsg dispatcher the following commands are added:

* orientationleft
* orientationtop
* orientationright
* orientationbottom
* orientationnext   (cycles clockwise)
* orientationprev  (cycles counter-clockwise)

Here's a screenshot of a vertical monitor using the bottom orientation, with the master area at the bottom, and two slave windows on top:

Screenshot of the bottom orientation

#### Is there anything you want to mention? (unchecked code, possible bugs, found problems, breaking compatibility, etc.)

Should be good to go.

#### Is it ready for merging, or does it need work?

Tested to the best of my ability and ready for merging

### CODE REVIEW

File: src/helpers/Workspace.hpp [hunk 1]
```diff
@@ -25,6 +34,7 @@ class CWorkspace {
     int             m_iPrevWorkspaceID = -1;
     bool            m_bHasFullscreenWindow = false;
     eFullscreenMode m_efFullscreenMode = FULLSCREEN_FULL;
+    eOrientation    m_orientation = ORIENTATION_LEFT;
```
Comment (vaxerski) [inline src/helpers/Workspace.hpp:37]: this has no place in `CWorkspace`, as it's a layout-specific thing.

Consider making a separate class / struct in MasterLayout for workspace-specific data, and storing it there.

e.g.
```cpp
struct stuff {
    int workspaceid = -1; 
    eorient orient = 0;
 }

...
    // in the layout, a private field
    std::vector<stuff> m_vStuff;
    stuff getStuffForWorkspace(int id);
```

File: src/layout/MasterLayout.cpp [hunk 2]
```diff
@@ -196,53 +196,120 @@ void CHyprMasterLayout::calculateWorkspace(const int& ws) {
...
 
+    //compute placement of master window(s)
     if (getNodesOnWorkspace(PWORKSPACE->m_iID) < 2) {
         PMASTERNODE->position = PMONITOR->vecReservedTopLeft + PMONITOR->vecPosition;
         PMASTERNODE->size = Vector2D(PMONITOR->vecSize.x - PMONITOR->vecReservedTopLeft.x - PMONITOR->vecReservedBottomRight.x, PMONITOR->vecSize.y - PMONITOR->vecReservedBottomRight.y - PMONITOR->vecReservedTopLeft.y);
         applyNodeDataToWindow(PMASTERNODE);
         return;
-    } else {
+
+    } else if ((PWORKSPACE->m_orientation == ORIENTATION_LEFT) || (PWORKSPACE->m_orientation == ORIENTATION_RIGHT))  {
```
Comment (vaxerski) [inline src/layout/MasterLayout.cpp:206]: useless parentheses around equalities

File: src/helpers/Workspace.hpp [hunk 3]
```diff
@@ -8,6 +8,15 @@ enum eFullscreenMode : uint8_t {
     FULLSCREEN_MAXIMIZED
 };
 
+//orientation is used by the Master layout
+//and determines which side of the screen the master area resides
+enum eOrientation: uint8_t {
+    ORIENTATION_LEFT =0,
```
Comment (vaxerski) [inline src/helpers/Workspace.hpp:14]: missing space before 0

File: src/layout/MasterLayout.cpp [hunk 4]
```diff
@@ -196,53 +196,120 @@ void CHyprMasterLayout::calculateWorkspace(const int& ws) {
...
+                n.size = Vector2D(WIDTH, HEIGHT);
 
                 nodesLeft--;
                 heightLeft -= HEIGHT;
                 nextY += HEIGHT;
 
+                applyNodeDataToWindow(&n);
+            }
+        }
+    } else if ((PWORKSPACE->m_orientation == ORIENTATION_TOP) || (PWORKSPACE->m_orientation == ORIENTATION_BOTTOM))  {
```
Comment (vaxerski) [inline src/layout/MasterLayout.cpp:231]: ditto

File: src/layout/MasterLayout.cpp [hunk 5]
```diff
@@ -196,53 +196,120 @@ void CHyprMasterLayout::calculateWorkspace(const int& ws) {
...
+        float widthLeft = PMONITOR->vecSize.x - PMONITOR->vecReservedBottomRight.x - PMONITOR->vecReservedTopLeft.x;
+        int nodesLeft = MASTERS;
+        float nextX = 0;
+        const float HEIGHT =  (PMONITOR->vecSize.y - PMONITOR->vecReservedBottomRight.y - PMONITOR->vecReservedTopLeft.y) * PMASTERNODE->percMaster;
+
+        for (auto& n : m_lMasterNodesData) {
+            if (n.workspaceID == PWORKSPACE->m_iID && n.isMaster) {
+                if (PWORKSPACE->m_orientation == ORIENTATION_BOTTOM)  {
```
Comment (vaxerski) [inline src/layout/MasterLayout.cpp:239]: double space before `{` - accident?

File: src/layout/MasterLayout.cpp [hunk 6]
```diff
@@ -196,53 +196,120 @@ void CHyprMasterLayout::calculateWorkspace(const int& ws) {
...
             }
         }
     }
 
-    float heightLeft = PMONITOR->vecSize.y - PMONITOR->vecReservedBottomRight.y - PMONITOR->vecReservedTopLeft.y;
+
+    //compute placement of slave window(s)
     int slavesLeft = getNodesOnWorkspace(PWORKSPACE->m_iID) - MASTERS;
-    float nextY = 0;
+    if ((PWORKSPACE->m_orientation == ORIENTATION_LEFT) || (PWORKSPACE->m_orientation == ORIENTATION_RIGHT)) {
```
Comment (vaxerski) [inline src/layout/MasterLayout.cpp:261]: ditto

File: src/layout/MasterLayout.cpp [hunk 7]
```diff
@@ -196,53 +196,120 @@ void CHyprMasterLayout::calculateWorkspace(const int& ws) {
...
+            slavesLeft--;
+            heightLeft -= HEIGHT;
+            nextY += HEIGHT;
 
-        slavesLeft--;
-        heightLeft -= HEIGHT;
-        nextY += HEIGHT;
+            applyNodeDataToWindow(&nd);
+        }
+    } else if ((PWORKSPACE->m_orientation == ORIENTATION_TOP) || (PWORKSPACE->m_orientation == ORIENTATION_BOTTOM)) {
```
Comment (vaxerski) [inline src/layout/MasterLayout.cpp:286]: ditto

File: src/layout/MasterLayout.cpp [hunk 8]
```diff
@@ -752,6 +819,61 @@ std::any CHyprMasterLayout::layoutMessage(SLayoutMessageHeader header, std::stri
             PNODE->isMaster = false;
         }
 
+        recalculateMonitor(header.pWindow->m_iMonitorID);
+    } else if ((message == "orientationleft") || (message == "orientationright") || (message == "orientationtop") || (message == "orientationbottom")) {
```
Comment (vaxerski) [inline src/layout/MasterLayout.cpp:823]: ditto

File: src/layout/MasterLayout.hpp [hunk 9]
```diff
@@ -5,6 +5,7 @@
 #include <deque>
 
 enum eFullscreenMode : uint8_t;
+enum eOrientation: uint8_t;
```
Comment (vaxerski) [inline src/layout/MasterLayout.hpp:8]: missing space before `:`

**Comment by vaxerski** [comment]
thanks! really cool. Left some minor code quality comments.

File: src/helpers/Workspace.hpp (hunk 1 above)
Comment (proycon) [reply src/helpers/Workspace.hpp:37]: Right, I too felt it was a bit misplaced there, I simply took the easiest path. I'll try improve it as you suggested.

Comment (proycon) [reply src/helpers/Workspace.hpp]: I implemented this now, one thing I'm still a bit unsure of is the on-the-fly struct creation I do in https://github.com/hyprwm/Hyprland/blob/444be966bca1022fc77c63e59a96528b1e4a8f1e/src/layout/MasterLayout.cpp#L40

File: src/layout/MasterLayout.cpp [hunk 10]
```diff
@@ -30,6 +30,28 @@ int CHyprMasterLayout::getMastersOnWorkspace(const int& ws) {
...
+SMasterWorkspaceData* CHyprMasterLayout::getMasterWorkspaceData(const int& ws) {
+    for (auto& n : m_lMasterWorkspacesData) {
+        if (n.workspaceID == ws)
+            return &n;
+    }
+
+    //create on the fly if it doesn't exist yet
+    const auto PWORKSPACEDATA = &m_lMasterWorkspacesData.emplace_back();
+    PWORKSPACEDATA->workspaceID = ws;
+    const auto orientation = g_pConfigManager->getString("master:orientation");
```
Comment (vaxerski) [inline src/layout/MasterLayout.cpp:42]: see https://wiki.hyprland.org/Contributing-and-Debugging/PR-Guidelines/#some-code-faq

Comment (proycon) [reply src/layout/MasterLayout.cpp]: Addressed now in 3d3431223181afc716ae46d7059dfc2ccadb5b5c , does this look better?

Comment (vaxerski) [reply src/layout/MasterLayout.cpp]: yes

**Comment by vaxerski** [comment]
I'll trust you that you tested it

**Comment by vaxerski** [comment]
Make a wiki MR with the new opt please inb4 I merge it

**Pull request merged**